### ISSUE TITLE
quasiquoter?

### ISSUE THREAD

**Issue by mwotton** [Porges/email-validate-hs#18]
it'd be nice to have a quasiquoter that can be run at compile time so i don't need to do (fromJust $ emailAddress "obviously@correct.com"). I had a bit of a go, but got a weird error from ByteString's Data.Data instance:

```
emailQQ = QuasiQuoter { quoteExp = quoteEmailExp
                      }

quoteEmailExp s = do
   case emailAddress (BS8.pack s) of
     Nothing -> error "bugger"
     Just x -> trace "worked" $ dataToExpQ (const Nothing) (x :: EmailAddress)
```

```
[1 of 1] Compiling Main             ( test/Spec.hs, .stack-work/dist/x86_64-linux/Cabal-1.22.5.0/build/email-validate-qq-test/email-validate-qq-test-tmp/Main.o )
worked       

/home/mark/projects/email-validate-qq/test/Spec.hs:12:12:
    Exception when trying to run compile-time code:
      Data.ByteString.ByteString.toConstr
    Code: template-haskell-2.10.0.0:Language.Haskell.TH.Quote.quoteExp
            emailQQ "aperson@arealaddress.com"
```

is the Data instance for ByteString just not up to it? I might be doing something terribly wrong.

**Comment by Porges** [comment]
I looked into this, it seems that the `ByteString` `Data` instance that `EmailAddress` inherits doesn't support `toConstr` which is called here... I'm not sure what to do in this case (should I even remove `Data`?)

I'll add an alternative implementation by 'manually' making the `Q Exp`. This works nicely.

**Comment by Porges** [comment]
Released in v2.3

**Comment by mwotton** [comment]
thanks :)
